### PR TITLE
Performance tuning

### PR DESCRIPTION
**Pull request by fmgjcoppens**
With these modifications I managed to shave off about 1/3 of the execution time (15 sec -> ~10 sec) on a single i9-10900 AVX2 core, without KOKKOS/MPI/OpenMP. Compiled LAMMPS with:

CXXFLAGS="-O2 -march=skylake -mfma -g -DNDEBUG" cmake ../cmake/ -D PKG_ML-UF3=yes -D BUILD_MPI=no -D PKG_KOKKOS=no -D BUILD_OMP=no

Running the 'lammps_plugin/tungsten_example/W_nve.in' input file.

### CODE REVIEW

**Comment by monk-04** [comment]
Thank you for your commits. I am running some test calculations on our system to measure the speed-up. I will post the result here

**Comment by monk-04** [comment]
I ran some test calculations on a system containing two elements and 64 atoms. The simulations were run using 1 Intel(R) Xeon(R) CPU E5-2698 v3 cpu on Hipergator at the University of Florida.

I used the GCC compiler version 9.3.0 to compile lammps without KOKKOS/MPI/OpenMP (`cmake ../cmake -D PKG_ML-UF3=yes -D BUILD_MPI=no -D PKG_KOKKOS=no -D BUILD_OMP=no`)

The performance is averaged over ten simulation runs, each consisting of 10000 MD steps at 2000K and 'nve' fix. 

Performance without optimization --> 5.8011 ns/day, 4.1372 hours/ns, 67.1426 timesteps/s
Performance with @fmgjcoppens optimization --> 7.5279 ns/day, 3.1961 hours/ns, 87.1298 timesteps/s

For the tungsten example -->
Performance without optimization --> 28.4621 ns/day, 0.8456 hours/ns, 329.4206 timesteps/s
Performance with @fmgjcoppens optimization --> 34.592 ns/day, 0.6938 hours/ns, 400.3688 timesteps/s

There is a noticeable improvement in the performance.

How about we use the `_alignvar` and `_noalias` macros defined in [lmphtype.h](https://github.com/lammps/lammps/blob/develop/src/lmptype.h) for the `__attribute` and `__restrict` macros to make them compiler independent?

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 1]
```diff
@@ -134,19 +134,19 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
 
   // Calculate energies
 
-  double basis_ij[4];
+  double __attribute__((aligned(8))) basis_ij[4]; // Force memory alignment on 8 byte boundaries
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double basis_ij[4], 8);  // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 2]
```diff
@@ -134,19 +134,19 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
   basis_ij[0] = bsplines_ij[iknot_ij].eval3(rth_ij, rsq_ij, value_rij);
   basis_ij[1] = bsplines_ij[iknot_ij + 1].eval2(rth_ij, rsq_ij, value_rij);
   basis_ij[2] = bsplines_ij[iknot_ij + 2].eval1(rth_ij, rsq_ij, value_rij);
   basis_ij[3] = bsplines_ij[iknot_ij + 3].eval0(rth_ij, rsq_ij, value_rij);
 
-  double basis_ik[4];
+  double __attribute__((aligned(8))) basis_ik[4]; // Force memory alignment on 8 byte boundaries
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double basis_ik[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 3]
```diff
@@ -134,19 +134,19 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
   basis_ik[0] = bsplines_ik[iknot_ik].eval3(rth_ik, rsq_ik, value_rik);
   basis_ik[1] = bsplines_ik[iknot_ik + 1].eval2(rth_ik, rsq_ik, value_rik);
   basis_ik[2] = bsplines_ik[iknot_ik + 2].eval1(rth_ik, rsq_ik, value_rik);
   basis_ik[3] = bsplines_ik[iknot_ik + 3].eval0(rth_ik, rsq_ik, value_rik);
 
-  double basis_jk[4];
+  double __attribute__((aligned(8))) basis_jk[4]; // Force memory alignment on 8 byte boundaries
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double basis_jk[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 4]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
 
   for (int i = 0; i < 4; i++) {
+    const double basis_iji = basis_ij[i]; // prevent repeated access of same memory location
     for (int j = 0; j < 4; j++) {
-      for (int k = 0; k < 4; k++) {
-        ret_val[0] += coeff_matrix[i + iknot_ij][j + iknot_ik][k + iknot_jk] * basis_ij[i] *
-            basis_ik[j] * basis_jk[k];
-      }
+      const double factor = basis_iji * basis_ik[j]; // prevent repeated access of same memory location
+      const double* __restrict__ slice = &coeff_matrix[i + iknot_ij][j + iknot_ik][iknot_jk]; // declare a contigues 1D slice of memory and let the compiler know that the memory that this pointer points to has no overlap with other pointers.
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: const double* _noalias slice = &coeff_matrix[i + iknot_ij][j + iknot_ik][iknot_jk];

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 5]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
+      tmp[3] = slice[3] * basis_jk[3];
+      double sum = tmp[0] + tmp[1] + tmp[2] + tmp[3];
+      ret_val[0] += factor * sum; // use 1 fused multiply-add (FMA)
     }
   }
 
   // Calculate forces
 
-  double dnbasis_ij[4];
+  double __attribute__((aligned(8))) dnbasis_ij[4];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double dnbasis_ij[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 6]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
   dnbasis_ij[0] = dnbsplines_ij[iknot_ij].eval2(rsq_ij, value_rij);
   dnbasis_ij[1] = dnbsplines_ij[iknot_ij + 1].eval1(rsq_ij, value_rij);
   dnbasis_ij[2] = dnbsplines_ij[iknot_ij + 2].eval0(rsq_ij, value_rij);
   dnbasis_ij[3] = 0;
 
-  double dnbasis_ik[4];
+  double __attribute__((aligned(8))) dnbasis_ik[4];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double dnbasis_ik[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 7]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
   dnbasis_ik[0] = dnbsplines_ik[iknot_ik].eval2(rsq_ik, value_rik);
   dnbasis_ik[1] = dnbsplines_ik[iknot_ik + 1].eval1(rsq_ik, value_rik);
   dnbasis_ik[2] = dnbsplines_ik[iknot_ik + 2].eval0(rsq_ik, value_rik);
   dnbasis_ik[3] = 0;
 
-  double dnbasis_jk[4];
+  double __attribute__((aligned(8))) dnbasis_jk[4];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double dnbasis_jk[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 8]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
 
   for (int i = 0; i < 3; i++) {
+    const double dnbasis_iji = dnbasis_ij[i];
     for (int j = 0; j < 4; j++) {
-      for (int k = 0; k < 4; k++) {
-        ret_val[1] += dncoeff_matrix_ij[iknot_ij + i][iknot_ik + j][iknot_jk + k] * dnbasis_ij[i] *
-            basis_ik[j] * basis_jk[k];
-      }
+      const double factor = dnbasis_iji * basis_ik[j];
+      const double* __restrict__ slice = &dncoeff_matrix_ij[iknot_ij + i][iknot_ik + j][iknot_jk];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: const double* _noalias slice = &dncoeff_matrix_ij[iknot_ij + i][iknot_ik + j][iknot_jk];

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 9]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
+      double __attribute__((aligned(8))) tmp[4];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double tmp[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 10]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
 
   for (int i = 0; i < 4; i++) {
+    const double basis_iji = basis_ij[i];
     for (int j = 0; j < 3; j++) {
-      for (int k = 0; k < 4; k++) {
-        ret_val[2] += dncoeff_matrix_ik[iknot_ij + i][iknot_ik + j][iknot_jk + k] * basis_ij[i] *
-            dnbasis_ik[j] * basis_jk[k];
-      }
+      const double factor = basis_iji * dnbasis_ik[j];
+      const double* __restrict__ slice = &dncoeff_matrix_ik[iknot_ij + i][iknot_ik + j][iknot_jk];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: const double* _noalias slice = &dncoeff_matrix_ik[iknot_ij + i][iknot_ik + j][iknot_jk];

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 11]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
+      double __attribute__((aligned(8))) tmp[4];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double tmp[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 12]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
 
   for (int i = 0; i < 4; i++) {
+    const double basis_iji = basis_ij[i];
     for (int j = 0; j < 4; j++) {
-      for (int k = 0; k < 3; k++) {
-        ret_val[3] += dncoeff_matrix_jk[iknot_ij + i][iknot_ik + j][iknot_jk + k] * basis_ij[i] *
-            basis_ik[j] * dnbasis_jk[k];
-      }
+      const double factor = basis_iji * basis_ik[j];
+      const double* __restrict__ slice = &dncoeff_matrix_jk[iknot_ij + i][iknot_ik + j][iknot_jk];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: const double* _noalias slice = &dncoeff_matrix_jk[iknot_ij + i][iknot_ik + j][iknot_jk];

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp [hunk 13]
```diff
@@ -158,58 +158,81 @@ double *uf3_triplet_bspline::eval(double value_rij, double value_rik, double val
...
+      double __attribute__((aligned(8))) tmp[3];
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: _alignvar(double tmp[3], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.h [hunk 14]
```diff
@@ -30,7 +30,8 @@ class uf3_triplet_bspline {
   std::vector<std::vector<double>> knot_matrix;
   std::vector<uf3_bspline_basis3> bsplines_ij, bsplines_ik, bsplines_jk;
   std::vector<uf3_bspline_basis2> dnbsplines_ij, dnbsplines_ik, dnbsplines_jk;
-  double ret_val[4];
+  
+  double __attribute__((aligned(8))) ret_val[4]; // Force memory alignment on 8 byte boundaries
```
Comment (monk-04) [inline lammps_plugin/ML-UF3/uf3_triplet_bspline.h]: _alignvar(double ret_val[4], 8); // Force memory alignment on 8 byte boundaries using lammps defined macros

File: lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp (hunk 1 above)
Comment (fmgjcoppens) [reply lammps_plugin/ML-UF3/uf3_triplet_bspline.cpp]: I switched the compiler dependent preprocessor macros __attribute__(( aligned(8))) and __restrict__ with the compile independent LAMMPS preprocessor macros _alignvar(x,y) and _noalias. 

For the tungsten example -->
commit  b4bf4d6 --> Performance: 85.892 ns/day, 0.279 hours/ns, 994.124 timesteps/s, 53.683 katom-step/s
commit  b300270 --> Performance: 87.461 ns/day, 0.274 hours/ns, 1012.275 timesteps/s, 54.663 katom-step/s

The difference is negligible.

**Comment by monk-04** [comment]
I ran some test calculations with [b300270](https://github.com/monk-04/uf3/pull/6/commits/b3002706aac4417fcb1400c71bd705831f4e0b06) commit. 

Performance --> 7.432 ns/day, 3.23 hours/ns, 86 timestep/s

There is a negligible difference in performance with compiler-independent macros. I think this pull request is ready to be merged.

**Review by monk-04 — APPROVED** [review]
Everything looks good to me!

**Pull request merged**